### PR TITLE
Add iteration spaces

### PR DESCRIPTION
**Pull request by hiker**
I think this branch is ready to be reviewed. I can now compile my "most" version without changes to PSyclone using `psyclone ... --config new_halos.psyclone ...` with new_halos.psyclone being:
```
[DEFAULT]
DISTRIBUTED_MEMORY = false
defaultstubapi=dynamo0.3

[gocean1.0]
iteration-spaces=offset_ne:ct:internal_we_halo:2:{stop}:1:{stop}+1
                 offset_ne:ct:internal_ns_halo:1:{stop}+1:2:{stop}
```
It includes a few (more or less ;) ) necessary changes:
1. If the config files does only contain one API section, this will become the default_api, so no need to specify `defaultapy` anymore in the case above (or have a `-api` command line parameter). 
2. I replaced the `ConfigFactory` as discussed in #191 and all global variables that stored the config instance. The side effect is that it appears to be possible to run now all tests in parallel (previously using -n with pytest resulted in many many failed tests, even on master).
3. I added 4 more GRID_PROPERTY_DICT to access minimum and maximum grid dimension (since my code tests to handle boundaries separately).
4. Added more test cases and fixed some style issues.

### CODE REVIEW

File: src/psyclone/tests/test_files/gocean1p0/new_iteration_space.f90 [hunk 1]
```diff
@@ -0,0 +1,53 @@
+!-------------------------------------------------------------------------------
+! (c) The copyright relating to this work is owned jointly by the Crown,
```
Comment (arporter) [inline src/psyclone/tests/test_files/gocean1p0/new_iteration_space.f90]: Please correct this copyright/license notice to be the current version.

File: src/psyclone/tests/gocean1p0_test.py [hunk 2]
```diff
@@ -1176,6 +1176,34 @@ def test05p1_kernel_invalid_iterates_over():
               api="gocean1.0")
 
 
+def test05p1_kernel_add_iteration_spaces():
+    '''Check that adding a new iteration space works
+    '''
+
+    # Add new iteration space 'dofs'
+    GOLoop.add_bounds("offset_sw:cu:dofs:1:2:3:{stop}")
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/gocean1p0_test.py]: test05p1_kernel_invalid_iterates_over tests this (and still passes).

File: src/psyclone/tests/config_test.py [hunk 3]
```diff
@@ -100,32 +107,37 @@ def int_entry(request):
...
+    # be executed. So to be certain explicitly delete the singleton here
+    # to force that the next line will test loading the default config file.
+    Config._instance = None
+    _config = Config.get()
+    assert isinstance(_config, Config)
+    # Check that we are creating a singleton instance
+    _config2 = Config.get()
+    assert _config is _config2
+
+    # Test that we an not create more than one instance
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/config_test.py]: pylint did not complain about this, so I left it the way it was.

File: src/psyclone/tests/gocean1p0_test.py [hunk 4]
```diff
@@ -43,6 +43,7 @@
 from psyclone.parse import parse
 from psyclone.psyGen import PSyFactory
 from psyclone.generator import GenerationError, ParseError
+from psyclone.gocean1p0 import GOLoop
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/gocean1p0_test.py]: Done

File: src/psyclone/tests/test_files/gocean1p0/new_iteration_space.f90 (hunk 1 above)
Comment (hiker) [reply src/psyclone/tests/test_files/gocean1p0/new_iteration_space.f90]: Done

File: src/psyclone/generator.py [hunk 5]
```diff
@@ -252,19 +253,16 @@ def main(args):
         choices=Profiler.SUPPORTED_OPTIONS,
         help="Add profiling hooks for either 'kernels' or 'invokes' even if a "
              "transformation script is used. Use at your own risk.")
-    parser.set_defaults(dist_mem=_CONFIG.distributed_memory)
+    parser.set_defaults(dist_mem=Config.get().distributed_memory)
 
+    parser.add_argument("--config", help="Additional config file with "
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/generator.py]: Fixed, and the connection to the env variable is explained in the docs now.

File: src/psyclone/gocean1p0.py [hunk 6]
```diff
@@ -404,86 +411,146 @@ def __init__(self, parent=None,
...
-        self._bounds_lookup = {}
         for grid_offset in SUPPORTED_OFFSETS:
-            self._bounds_lookup[grid_offset] = {}
+            GOLoop._bounds_lookup[grid_offset] = {}
             for gridpt_type in VALID_FIELD_GRID_TYPES:
-                self._bounds_lookup[grid_offset][gridpt_type] = {}
+                GOLoop._bounds_lookup[grid_offset][gridpt_type] = {}
                 for itspace in VALID_ITERATES_OVER:
-                    self._bounds_lookup[grid_offset][gridpt_type][itspace] = {}
+                    GOLoop._bounds_lookup[grid_offset][gridpt_type][\
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/gocean1p0.py]: Yes, but otherwise pep8 complained ... but I finally found a way to make them both happy :)

File: src/psyclone/gocean1p0.py [hunk 7]
```diff
@@ -404,86 +411,146 @@ def __init__(self, parent=None,
...
+
+        # Check field-type exists
+        if not data[1] in current_bounds[data[0]]:
+            current_bounds[data[0]][data[1]] = {}
+
+        # Check iteration space exists:
+        if not data[2] in current_bounds[data[0]][data[1]]:
+            current_bounds[data[0]][data[1]][data[2]] = {}
+            VALID_ITERATES_OVER.append(data[2])
+        current_bounds[data[0]][data[1]][data[2]] = \
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/gocean1p0.py]: I can't really test the offset and field types, since I want to be able to support new field types etc (i.e. different things than just dl_esm_inf), so I left this open on purpose.
For start and stop - that's difficult. I've added a test that all "{...} expressions are "{start}" and "{stop"} only, but otherwise I am not sure what to check - maybe a different infrastructure library will have a value for halo size, so the start index could be something like: "{start}-halo_size" (halo size being somehow set depending on the field). I could try to parse the expression, but I am not sure if this is really worth the effort?

File: src/psyclone/gocean1p0.py [hunk 8]
```diff
@@ -404,86 +411,146 @@ def __init__(self, parent=None,
...
+        bound_info = offset_ne:ct:all_pts:outer:1:{stop}+1:2:{stop}
+
+        :param str bound_info: A string that contains a ":" separated \
+               tuple with the iteration space definition.
+        :raises ValueError: if bound_info is not a string.
+        :raises ConfigurationError: if bound_info is not formatted correctly.
+        '''
+
+        if not isinstance(bound_info, str):
+            raise ValueError("The parameter 'bound_info' must be a string, "
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/gocean1p0.py]: Done

File: src/psyclone/configuration.py [hunk 9]
```diff
@@ -164,11 +176,20 @@ def __init__(self, config_file=None):
...
+        else:
+            # Test if we have exactly one section (besides DEFAULT).
+            # If so, make this section the default API
+            if len(self._config) == 2:
+                for section in self._config:
+                    self._default_api = section.lower()
+                    if self._default_api != "default":
+                        break
+            else:
+                raise ConfigurationError("No DEFAULTAPI specified.",
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: All done.

**Comment by hiker** [comment]
OK, I think I have addresses the issue raised. I first tried to only parse a 'do i=1,n' statement - but parsing this alone more or less accepted everything (e.g. 'do i=1),2+' would be accepted). When I switched to use a full do statement ('do i=1, 2  \n a=1 \n end do") I got proper error handling. See https://github.com/stfc/PSyclone/blob/056012b007da6b6c4e3d746326f37d3988778fd7/src/psyclone/gocean1p0.py#L559-L576
I likely did something wrong, but not sure what. If you don't like parsing a full 'do' block, can you point out to me how to get errors reported for a 'do' statement only (or I guess I could even test an assignment instead of a full do loop even).

One other comment why I am not that concerned about correctness: I envision that the config file will be supplied by the infrastructure library (e.g. dl_esm_inf), and would have been tested to be correct. The long term goal being to be able to use a different infrastructure library without changes to PSyclone (and each infrastructure library providing its own config file). Now I am aware that this is a very difficult goal (not even sure if it is possible ;) ), and I won't focus on it. But this is a step in this direction, and it allows me to do any further work on _most_ without having to modify PSyclone, which is nice :)

Back to review 😃

**Comment by TeranIvy** [comment]
Hi @hiker, I think you have just unintentionally found a bug in fparser2 :-) 

`DO` statement is a child of the `Block_Nonlabel_Do_Construct` and can be accessed via `content` of the parent `BlockBase`, e.g. if 
```
  obj = Block_Nonlabel_Do_Construct(get_reader('''\
      do i=1,10
        a = 1
      end do
    '''))
```
then `obj.content[0]` will give you `do i=1,10` which is the instance of `Nonlabel_Do_Stmt` class which you would like to test. To test it directly just do ` _ = Nonlabel_Do_Stmt(get_reader(do_stmt))`.

`Nonlabel_Do_Stmt` in turn searches for `DO` keyword and everything else in the statement goes under the `Loop_Control` class, which is the place where incorrect loop counters/indices will be caught.

The incorrect Fortran in your example (`do i=1),2+`) will get caught, so this is good news. However, I just played with trying out `do i=1, 2, z` and this did not get caught - so I will raise an issue to fix `Loop_Control` class in fparser2.

I hope this helps!

**Comment by hiker** [comment]
Hi @TeranIvy,

thanks for the answer ... though I admit I am somewhat confused 😕 
> Hi @hiker, I think you have just unintentionally found a bug in fparser2 :-)

That I get :)
...
> then `obj.content[0]` will give you `do i=1,10` which is the instance of `Nonlabel_Do_Stmt` class which you would like to test. To test it directly just do `_ = Nonlabel_Do_Stmt(get_reader(do_stmt))`.

I tried this with a code like this:
```
do_loop = "do i=1), 2+"
try:
    x = Nonlabel_Do_Stmt(get_reader(do_loop))
except Exception as err:
```
and this does not raise an exception. OK, so this would be the bug you mentioned I guess??
...
> The incorrect Fortran in your example (`do i=1),2+`) will get caught, 

With using Nonlabel_Do_Stmt? This appears not to be the case for me. I assume it should raise an exception? Or do I need to test its return value, or check if all tokens have been consumed or so?

> so this is good news. However, I just played with trying out `do i=1, 2, z` and this did not get caught - so I will raise an issue to fix `Loop_Control` class in fparser2.

Your example is actually syntactically correct Fortran, so here is where I am finally confused ;)

Even stranger:
```
do_loop="i=1("
_ = Scalar_int_expr(get_reader(do_loop))
```
(and I admit that this was a typo originally) - this did pass all unit tests, so it must trigger the right exceptions. So parsing an assignment to a variable as Scalar_int_expr seems to work just fine - which is at least odd (since it works with valid boundaries, and raises an exception with the invalid boundary values)

But the basic summary seems to be: for now I leave the test the way it is till fparser is fixed - which I can easily do :)

**Comment by TeranIvy** [comment]
Hi @hiker,
> Your example is actually syntactically correct Fortran, so here is where I am finally confused ;)

You're right, I always forget that (combinations of) letters can be used for numbers when trying to think of fparser2 examples to add - it's called variable declaration :-).

I think I found what the problem is in fparser2. I only ran your examples by using `Nonlabel_Do_Stmt` directly on the string
```
do_loop = "do i=1), 2+"
objt = Nonlabel_Do_Stmt(do_loop)
```
which falls over nicely as it should with `NoMatchError: Nonlabel_Do_Stmt: 'do i=1), 2+'`.

In your example `get_reader(do_stmt)` returns `None` so running `objt = Nonlabel_Do_Stmt(get_reader(do_loop))` returns `NoneType` instead of raising an error. It seems that `Loop_Control` needs fixing.

**Comment by hiker** [comment]
Hi @TeranIvy ,

ah - without get_reader(). I didn't know that this is possible ;) But that works for me. Now I only parse a do statement (not the whole block) - thanks for your help!

Ready for review :)

File: doc/configuration.rst [hunk 10]
```diff
@@ -79,27 +85,27 @@ supported by PSyclone.
...
 REPRODUCIBLE_REDUCTIONS Whether or not to generate code for reproducible OpenMP
                         reductions (see :ref:`openmp-reductions`) by default.
 REPROD_PAD_SIZE         If generating code for reproducible OpenMP reductions,
                         this setting controls the amount of padding used
-			between elements of the array in which each thread
-			accumulates its local reduction. (This prevents false
+                        between elements of the array in which each thread
+                        accumulates its local reduction. (This prevents false
                         sharing of cache lines by different threads.)
 ======================= =======================================================
```
[the remark this answers is not in the document]
Comment (hiker) [reply doc/configuration.rst]: One suggestion that might make the situation a bit more convenient: what about instead of putting DEFAULTAPI into the config file, we only put API in the config file. Then we can say that we have a default, can print it in the help messages, and still maintain the functionality we have (with this patch) ... as far as I can see.

File: src/psyclone/gocean1p0.py [hunk 11]
```diff
@@ -404,86 +411,194 @@ def __init__(self, parent=None,
...
+        if len(data) != 7:
+            from psyclone.configuration import ConfigurationError
+            raise ConfigurationError("An iteration space must be in the form "
+                                     "\"offset-type:field-type:"
+                                     "iteration-space:outer-start:"
+                                     "outer-stop:inner-start:inner-stop\"\n"
+                                     "But got \"{0}\"".format(bound_info))
+
+        if len(GOLoop._bounds_lookup) == 0:
+            GOLoop.setup_bounds()
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/gocean1p0.py]: Strange, for me it is reported to be covered. I modified gocen1p0_config_test.py so that only the first test would run, the one which includes:
```
   # Check that --config with a parameter is accepted
    main(options+["--config", config_file, f90_file])
```
joerg@localhost:~/work/psyclone/src/psyclone/tests> py.test  --cov-report=html  --cov=psyclone.gocean1p0  --cov-report term-missing  ./gocean1p0_config_test.py 
========================================================================================== test session starts ===========================================================================================
platform linux2 -- Python 2.7.14, pytest-3.6.1, py-1.5.3, pluggy-0.6.0
rootdir: /home/joerg/work/psyclone, inifile:
plugins: xdist-1.23.0, pylint-0.7.1, pep257-0.0.5, pep8-1.0.6, forked-0.2, flakes-2.0.0, cov-2.5.1
collected 1 item                                                                                                                                                                                         

gocean1p0_config_test.py .                                                                                                                                                                         [100%]

---------- coverage: platform linux2, python 2.7.14-final-0 ----------
Name                                                  Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------
/home/joerg/work/psyclone/src/psyclone/gocean1p0.py     535    157    71%   144, 164, 194, 218, 231, 265-268, 271-274, 318-322, 326-331, 337-338, 350, 364, 378, 411, 415, 520, 526-527, 545-546, 568-570, 580, 584, 633-664, 707-728, 734-744, 753, 760, 765, 774, 796, 841, 862, 881, 908, 915, 941, 956, 971, 993-1017, 1028-1029, 1041-1042, 1076, 1112, 1122, 1194, 1206-1263, 1268, 1286-1287, 1302-1303, 1313-1314, 1342-1348, 1391, 1406-1413, 1421, 1430, 1452, 1464, 1469, 1477, 1482, 1493, 1515, 1555-1564

This does not list line 533 or 534 as uncovered :( I didn't check if this is correct - but it again raises the question: which version are you using? Maybe I need to update, or change the parameters I am using for testing?

**Comment by hiker** [comment]
Most of the changes done, three issues/question:
1) I had to fix some fparser imports (walk_ask for me was moved from Fparser2003 into utils) - now it fails here, so I am now confused - I checked, travis reports "4c4cc6a8c5433658e62321eab34c9804b3301d00", which is the same I have ... why on earth is walk_ast in a different location??
2) I did move the Config._instance to the tearDown function - though I want to raise the potential issue that in the first test that is being run this is not done anymore. That means that depending on what the previous test from a different test-file did, and what the first test in those files is now, it will get a potentially unknown/invalid/... config file. Previously we had a reset before any of those tests, now we have a reset only after. So far it seems to work, and might be a non-issue in reality ... but who knows. TBH, I preferred the previous way of explicitly removing the config files for any tests that needed to start from scratch. 
3) I need an answer re defaultapi. My suggestion: we rename 'defaultapi' in the config file to 'api', rename the api function to api_config everywhere (which returns the api-specific config file, which I think is the better name than just 'api', and add a (then) new function 'api', which is the api selected by the user). Then we add a default_api to the Config file, hard-coding it to dynamo0.3. The result of this is: we have a default api, which we can show in help messages etc. The user can specify the API in either the config file or the command line (command line will then set the api definition in the Config class). The 5 or so usages of Config.get().default_api will be changed to Config.get().api, so in any function that loads the (previously default_api) in case it does not get an API, it will now load the user selected API (and if the user does not specify an API, it will still be using the default_api as api.

I think 3) will give a better overall handling of the API: default can be shown in help message, user can specify api either via command line or config file. So ... here the thread: if I don't hear from you, I'll do it this way ;)

**Comment by hiker** [comment]
OK, I think I solved the first issue (and corresponding travis failures): after switching to branches backwards and forwards, I had apparently stale pyc files in fparser (with no corresponding py file). This explain why I had to apply 4ef5f9f2104d37121638c67c456e067d51abe75d in order to get all tests to work. I reverted this in a10b19fb8fab1165180a5040d878133b51cf6415 ... and hope that it all works now.

So, now I am mostly waiting on comments on the default_api changes 😺

File: src/psyclone/gocean1p0.py [hunk 12]
```diff
@@ -404,86 +411,194 @@ def __init__(self, parent=None,
             raise GenerationError(
                 "Invalid loop type of '{0}'. Expected one of {1}".
                 format(self._loop_type, VALID_LOOP_TYPES))
+        if len(GOLoop._bounds_lookup) == 0:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/gocean1p0.py]: $ pylint --version
    pylint 1.6.4, 
    astroid 1.4.8
    Python 2.7.13 (default, Dec  1 2017, 09:21:53)

File: doc/configuration.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/configuration.rst]: Oh, I like that. So we put a default in e.g. `psyGen.py` but overwrite that if API is in the config file? If that's what you're proposing then I'm happy with that.

File: src/psyclone/gocean1p0.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/gocean1p0.py]: That is strange. I (almost) always work on what codecov says as that way I don't have to do it myself :-) Click on 'Continue to review full report at Codecov.` in the codecov comment on the PR and go to the `Diff` tab.

File: doc/configuration.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (hiker) [reply doc/configuration.rst]: Mostly, except I would put the API default in Config._default_api (so that all configuration items are in one place), and then in psyGen.get_api() use:
```
    if api == "":
        api = Config.get().api    # instead of api = Config.get().default_api
```
If the user does not set an API (neither as parameter to psyGen.get_api(), nor on the command line or config file), then the Config._api will be set to Config._default_api, so get_api() will then return the default_api.

**Comment by arporter** [comment]
re. the `walk_ast` - that has moved from the `Fortran2003` to the `utils` module as part of the re-structuring work being done on fparser2. Apologies for the confusion. I sometimes hit problems when swapping between branches and have to do a `git submodule update` to get the correct fparser branch. I'm not sure why this is - I'd have assumed that git was smart enough to get this right when changing branches so maybe I too have issues with stale pyc files.

File: doc/configuration.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/configuration.rst]: Oh yes - ignore me! I was forgetting we have a Config class for that now :-)

**Comment by arporter** [comment]
How about we change the `teardown` to a `setup` and wipe the config object in there instead?

File: src/psyclone/gocean1p0.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/gocean1p0.py]: OK, I think I found the reason for it: my command line runs only a few of the tests (e.g. only the gocean1p0 and the config_test - but when I run all tests I get the same failure (due to the runtime on my VM I usually don't do that). OK, I'll figure out how to reset the test in gocean1p0_config_test.py appropriately so that it will always test that line.

**Comment by arporter** [comment]
Hi @hiker, I've lost track of where this PR is up to. Is it in need of a review now?

**Comment by hiker** [comment]
Hi,

no, I am finishing it, but I had some higher priority other stuff to do
at work, and then some leave :)  Back to it tomorrow, though it will
probably be another day or two - sorry for the delay.

Cheers,
   Joerg

File: doc/configuration.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (hiker) [reply doc/configuration.rst]: OK, that should all be done now. One very minor question: some classes take an api as an optional argument (psyGen.get_api, parse.KernelTypeFactory.__init__, parse.parse), and if none is specified, they use Config.get().default_api. It might make sense to use Config.get().api (which is the same as default_api if no api is specified on the command line or a config file anyway) instead ... but I don't think it actually matters :)

File: src/psyclone/gocean1p0.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/gocean1p0.py]: OK, I am on 1.6.5/1.4.9/2.7.14. Anyway, I fixed those issues, even though my pylint seems to be ok with it.

File: src/psyclone/configuration.py [hunk 13]
```diff
@@ -66,61 +66,68 @@ def __str__(self):
...
 
-    :param str config_file: Override default configuration file to read.
-    :raises ConfigurationError: if there are errors or inconsistencies in \
+        if Config._instance is not None:
+            raise ConfigurationError("Only one instance of "
+                                     "Config can be created")
+
+        # Setup the list of supported APIs and stubs before reading any
+        # config file:
+        self._supported_api_list = ["gunghoproto", "dynamo0.1", "dynamo0.3",
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: I am not sure how to reference it, and where to actually in that section. But the declaration has been moved, and the rst files have the name and list of APIs in it ... so I'll resolve this.

File: src/psyclone/tests/config_test.py [hunk 14]
```diff
@@ -210,107 +223,110 @@ def test_read_values():
...
+    config = Config()
+    config.load(config_file=TEST_CONFIG)
     # Check the setter method
-    _config.distributed_memory = False
-    assert not _config.distributed_memory
+    config.distributed_memory = False
+    assert not config.distributed_memory
     with pytest.raises(ConfigurationError) as err:
-        _config.distributed_memory = "not-a-bool"
+        # pylint: disable=redefined-variable-type
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/config_test.py]: Mine does not, so I'll leave everything the way it is.

**Comment by hiker** [comment]
The last 'minor' changes were a lot more involved. But I think I have addressed all issues, ready for another review.

File: src/psyclone/configuration.py [hunk 15]
```diff
@@ -107,13 +112,10 @@ def __init__(self):
 
         # Setup the list of supported APIs and stubs before reading any
         # config file:
-        self._supported_api_list = ["gunghoproto", "dynamo0.1", "dynamo0.3",
-                                    "gocean0.1", "gocean1.0"]
-        self._supported_stub_api_list = ["dynamo0.3"]
-        self._api = None
+        self._api_conf = None
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/configuration.py]: In fact, please could you do that for each of the properties being initialised/declared here.

File: src/psyclone/gen_kernel_stub.py [hunk 16]
```diff
@@ -65,13 +62,13 @@ def generate(filename, api=""):
        format.
     '''
     if api == "":
```
Comment (arporter) [inline src/psyclone/gen_kernel_stub.py]: Please document the arguments to this routine in the docstring.

File: src/psyclone/transformations.py [hunk 17]
```diff
@@ -1013,7 +1009,7 @@ def apply(self, node, reprod=None):
         '''
 
         if reprod is None:
```
Comment (arporter) [inline src/psyclone/transformations.py]: Please document the arguments to the method (as you've touched it :-( )

File: doc/developers.rst [hunk 18]
```diff
@@ -1023,7 +1023,7 @@ The full interface to each of these classes is detailed below:
     :members:
     :noindex:
 
-configuration
+Configuration
```
[the remark this answers is not in the document]
Comment (hiker) [reply doc/developers.rst]: I so do not agree with that :)  As a compromise, what about using:
Module: configuration
?

File: src/psyclone/configuration.py [hunk 19]
```diff
@@ -171,45 +173,42 @@ def load(self, config_file=None):
...
-                                         config=self)
 
         # Sanity check
-        if self._default_api not in self._supported_api_list:
+        if self._api not in Config._supported_api_list:
             raise ConfigurationError(
-                "The default API ({0}) is not in the list of supported "
-                "APIs ({1}).".format(self._default_api,
-                                     self._supported_api_list),
+                "The API ({0}) is not in the list of supported "
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/configuration.py]: That's already there as well, line 196 :)

**Comment by hiker** [comment]
I think all done, and I hope you agree with using "Module: transformations' etc.

File: doc/developers.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/developers.rst]: OK :-)

File: src/psyclone/configuration.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/configuration.py]: So it is. I think I must have got confused as to which version I was looking at on github :-(
